cobigen-eclipse/cobigen-eclipse/src/com/devonfw/cobigen/eclipse/generator/GeneratorWrapperFactory.java

```diff
@@ -207,40 +212,78 @@ else if (selection instanceof IStructuredSelection) {
-  private static CobiGen initializeGenerator() throws InvalidConfigurationException, GeneratorCreationException {
+  private static CobiGen initializeGenerator(boolean allowMonolithicConfiguration)
+      throws InvalidConfigurationException, GeneratorCreationException {
 
     try {
       ResourcesPluginUtil.refreshConfigurationProject();
-      IProject generatorProj = ResourcesPluginUtil.getGeneratorConfigurationProject();
+      generatorProj = ResourcesPluginUtil.getGeneratorConfigurationProject();
 
-      if (generatorProj == null) {
-        throw new GeneratorCreationException(
-            "Configuration source could not be read. Have you downloaded the templates?");
+      initializeCobiGen(generatorProj, allowMonolithicConfiguration);
+
+      return initializeCobiGen(generatorProj, true);
+    } catch (CoreException e) {
+      throw new GeneratorCreationException("An eclipse internal exception occurred", e);
+    } catch (DeprecatedMonolithicConfigurationException e) {
+      if (null == generatorProj.getLocationURI())
+        throw e;
+      else {
+        throw new DeprecatedMonolithicConfigurationException(Paths.get(generatorProj.getLocationURI()));
       }
+    } catch (Throwable e) {
+      throw new GeneratorCreationException(
+          "Configuration source could not be read.\nIf you were updating templates, it may mean"
+              + " that you have no internet connection,",
+          e);
+    }
+  }
 
-      // We need to check whether it is a valid Java Project
-      IJavaProject configJavaProject = JavaCore.create(generatorProj);
+  /**
+   * Initializes the {@link CobiGen} with the correct configuration, checks if template-sets exists, if not the Jar or
+   * generatorProj will be used.
+   *
+   * @param generatorProj the templates project
+   * @param allowMonolithicConfiguration ignores deprecated monolithic template folder structure and if found does not
+   *        throw a DeprecatedMonolithicConfigurationException
+   * @return the configured{@link CobiGen}
+   * @throws if the generator configuration project does not exist
+   *
+   */
+  private static CobiGen initializeCobiGen(IProject generatorProj, boolean allowMonolithicConfiguration)
+      throws GeneratorCreationException {
 
-      // If it is not valid, we should use the jar
-      if (null == generatorProj.getLocationURI() || !configJavaProject.exists()) {
-        File templatesDirectory = CobiGenPaths.getTemplatesFolderPath().toFile();
-        File jarPath = TemplatesJarUtil.getJarFile(false, templatesDirectory);
-        boolean fileExists = jarPath.exists();
+    Path templatesDirectoryPath = CobiGenPaths.getTemplateSetsFolderPath();
+    Path templateSetsAdaptedFolderPath = templatesDirectoryPath.resolve(ConfigurationConstants.ADAPTED_FOLDER);
+    Path templateSetsDownloadedFolderPath = templatesDirectoryPath.resolve(ConfigurationConstants.DOWNLOADED_FOLDER);
+
+    if (ResourcesPluginUtil.getTemplateSetPathAfterUpgrade() != null) {
```

Comment:
Should the upgrader always be startet? We should also check the allowMonolithicConfiguration param.